haystack/agents/base.py

```diff
@@ -0,0 +1,298 @@
+        # if result was returned by a pipeline it is of type dict that we can use directly.
+        if isinstance(result, tuple):
+            result = result[0]
+        if isinstance(result, dict):
+            if "results" in result and len(result["results"]) > 0:
+                observation = result["results"][0]
+            elif "answers" in result and len(result["answers"]) > 0:
+                observation = result["answers"][0].answer
+            elif "documents" in result and len(result["documents"]) > 0:
+                observation = result["documents"][0].content
+            else:
+                # no result/answer/document was returned
+                observation = ""
+        return observation
+
+    def _extract_tool_name_and_tool_input(self, pred: str) -> Tuple[Optional[str], Optional[str]]:
+        """
+        Parse the tool name and the tool input from the prediction output of the Agent's PromptNode.
+
+        :param pred: Prediction output of the Agent's PromptNode from which to parse the tool and tool input
+        """
+
+        tool_match = re.search(self.tool_pattern, pred)
+        if tool_match:
+            tool_name = tool_match.group(1)
+            tool_input = tool_match.group(3)
+            return tool_name.strip('" []').strip(), tool_input.strip('" ')
+        return None, None
+
+    def _extract_final_answer(self, pred: str) -> Optional[str]:
+        """
+        Parse the final answer from the prediction output of the Agent's PromptNode.
+
+        :param pred: Prediction output of the Agent's PromptNode from which to parse the final answer
+        """
+        final_answer_match = re.search(self.final_answer_pattern, pred)
+        if final_answer_match:
+            final_answer = final_answer_match.group(1)
+            return final_answer.strip('" ')
+        return None
+
+    def _format_answer(
+        self, query: str, answer: str, tool_names: str, tool_names_with_descriptions: str, generated_text: str
+    ) -> Dict[str, Any]:
+        """
+        Formats an answer as a dict containing `query` and `answers` similar to the output of a Pipeline.
+        The full transcript based on the Agent's initial prompt template and the text it generated during execution.
+
+        :param query: The search query.
+        :param answer: The final answer returned by the Agent. An empty string corresponds to no answer.
+        :param tool_names: The names of the tools that will be inserted into the template.
+        :param tool_names_with_descriptions: The tool descriptions that will be inserted into the template.
+        :param generated_text: The text that the Agent generated in addition to the initial prompt template.
+        """
+        return {
+            "query": query,
+            "answers": [Answer(answer=answer, type="generative")],
+            "transcript": next(
+                self.prompt_node.get_prompt_template(self.prompt_template).fill(
```

Comment:
We have a reference to the template in `self.prompt_template`. Can we call it directly or it has to be via prompt node?

Reply:
`self.prompt_template` is just the name of the template. So to get the `think-step-by-step` template I think we always need the prompt node.

Reply:
I'll change the implementation so that `self.prompt_template` is a `PromptTemplate`. The function call will become `next(self.prompt_template.fill(` instead of `next(self.prompt_node.get_prompt_template(self.prompt_template).fill(` instead of